### ISSUE TITLE
BBC: Blead Breaks Mojolicious

### ISSUE THREAD

**Issue by cjg-cguevara** [Perl/perl5#22504]
This is a bug report for perl from "Carlos Guevara" <carlos@carlosguevara.com>,
generated with the help of perlbug 1.43 running under perl 5.41.3.

-----------------------------------------------------------------

BBC: Blead Breaks Mojolicious

Please see http://fast-matrix.cpantesters.org/?dist=Mojolicious

---
**Flags**
- category=core
- severity=low
---
**Perl configuration**
```
Site configuration information for perl 5.41.3:

Configured by cpan at Mon Aug 12 22:33:43 EDT 2024.

Summary of my perl5 (revision 5 version 41 subversion 3) configuration:
  Commit id: c2cb8f95a03c311972b110fc1fb6029c46446cad
  Platform:
    osname=linux
    osvers=5.14.0-427.28.1.el9_4.x86_64
    archname=x86_64-linux
    uname='linux cjg-rhel9 5.14.0-427.28.1.el9_4.x86_64 #1 smp preempt_dynamic fri jul 19 14:40:47 edt 2024 x86_64 x86_64 x86_64 gnulinux '
    config_args='-des -Dprefix=/home/cpan/bin/perl -Dscriptdir=/home/cpan/bin/perl/bin -Dusedevel -Duse64bitall'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=undef
    usemultiplicity=undef
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='cc'
    ccflags ='-fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2'
    optimize='-O2'
    cppflags='-fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='11.4.1 20231218 (Red Hat 11.4.1-3)'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib /usr/lib64 /usr/local/lib64
    libs=-lpthread -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -ldl -lm -lcrypt -lutil -lc
    libc=/lib/../lib64/libc.so.6
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version='2.34'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E'
    cccdlflags='-fPIC'
    lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector-strong'


---
@INC for perl 5.41.3:
    /home/cpan/bin/perl/lib/site_perl/5.41.3/x86_64-linux
    /home/cpan/bin/perl/lib/site_perl/5.41.3
    /home/cpan/bin/perl/lib/5.41.3/x86_64-linux
    /home/cpan/bin/perl/lib/5.41.3

---
Environment for perl 5.41.3:
    HOME=/home/cpan
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LC_ALL=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/cpan/bin/perl/bin:/home/cpan/bin:/usr/share/Modules/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash
```

**Comment by mauke** [comment]
The first error is:
```
Can't locate Mojo/BaseTestTest.pm in @INC (you may need to install the Mojo::BaseTestTest module) (@INC entries checked: /home/cpan/.cpan/build/Mojolicious-9.37-0/t/mojo/lib /home/cpan/.cpan/build/Mojolicious-9.37-0/blib/lib /home/cpan/.cpan/build/Mojolicious-9.37-0/blib/arch /home/cpan/bin/perl/lib/site_perl/5.41.3/x86_64-linux-thread-multi /home/cpan/bin/perl/lib/site_perl/5.41.3 /home/cpan/bin/perl/lib/5.41.3/x86_64-linux-thread-multi /home/cpan/bin/perl/lib/5.41.3 .) at /home/cpan/.cpan/build/Mojolicious-9.37-0/blib/lib/Mojo/Base.pm line 132.
BEGIN failed--compilation aborted at t/mojo/base.t line 22.
t/mojo/base.t .............................. 
Dubious, test returned 2 (wstat 512, 0x200)
No subtests run 
```
The second error is:
```

    #   Failed test 'loaded successfully'
    #   at t/mojo/loader.t line 79.

    #   Failed test 'loaded successfully'
    #   at t/mojo/loader.t line 80.
    # Looks like you failed 2 tests of 4.

#   Failed test 'Load'
#   at t/mojo/loader.t line 84.
# Looks like you failed 1 test of 15.
t/mojo/loader.t ............................ 
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/15 subtests 
```

The corresponding lines from t/mojo/base.t and t/mojo/loader.t are
```perl
use Mojo::Base "Mojo'BaseTestTest";
```
and
```perl
  ok !load_class("Mojo'LoaderTest::A"), 'loaded successfully';
```
respectively, so this is almost certainly the (expected) result of removing `'` as a package separator. Which would be commit ce1383eae4cd16.

**Comment by jkeenan** [comment]
Confirmed.  Bisection with the following invocation:
```
perl Porting/bisect.pl \
--module=Mojolicious \
--start=2463f19365f941f68e9d5eed2f787341df8ccdef \
--end=c2cb8f95a03c311972b110fc1fb6029c46446cad
```
... points to ce1383eae4 
```
commit ce1383eae4cd16b18ff9df6cecb54582e0c689ab
Author:     Tony Cook <tony@develop-help.com>
AuthorDate: Wed Jun 19 15:33:01 2024 +1000
Commit:     Tony Cook <tony@develop-help.com>
CommitDate: Mon Aug 12 10:13:37 2024 +1000

    remove use of ' as a package separator
```

v5.41.2-243-gce1383eae4

Notified upstream in https://github.com/mojolicious/mojo/issues/2194; also: https://github.com/mojolicious/mojo/pull/2193

**Comment by jkeenan** [comment]
> 
> Notified upstream in [mojolicious/mojo#2194](https://github.com/mojolicious/mojo/issues/2194); also: [mojolicious/mojo#2193](https://github.com/mojolicious/mojo/pull/2193)

The Mojolicious folks got their mojo working!  [Fixed in version 9.38 released Aug 17 2024](https://metacpan.org/dist/Mojolicious/changes).
